engine/src/test/java/com/datatorrent/stram/cli/ApexCliShutdownCommandTest.java

```diff
@@ -0,0 +1,142 @@
+import static org.powermock.api.mockito.PowerMockito.when;
+import static org.powermock.api.support.membermodification.MemberMatcher.constructor;
+import static org.powermock.api.support.membermodification.MemberModifier.suppress;
+
+/**
+ *
+ */
+@PowerMockIgnore({
+    "javax.net.ssl.*",
+    "org.apache.log4j.*"
+    })
+@PrepareForTest({YarnClient.class, ApexCli.class, StramAgent.class})
+public class ApexCliShutdownCommandTest
+{
+
+  @Rule
+  public PowerMockRule powerMockRule = new PowerMockRule();
+
+  private ApplicationReport mockRunningApplicationReport(String appId, String appName)
+  {
+    ApplicationReport app = mock(ApplicationReport.class);
+    ApplicationId applicationId = mock(ApplicationId.class);
+
+    when(applicationId.toString()).thenReturn(appId);
+    when(app.getApplicationId()).thenReturn(applicationId);
+
+    when(app.getName()).thenReturn(appName);
+
+    when(app.getYarnApplicationState()).thenReturn(YarnApplicationState.RUNNING);
+    when(app.getFinalApplicationStatus()).thenReturn(FinalApplicationStatus.UNDEFINED);
+
+    when(app.getTrackingUrl()).thenReturn("http://example.com");
+
+    return app;
+  }
+
+  @Test
+  public void shutdownAppCommandUsesBestEffortApproach() throws Exception
+  {
+    // Given a cli and two running apps
+    ApexCli cliUnderTest = new ApexCli();
+    StramAgent stramAgent = mock(StramAgent.class);
+    YarnClient yarnClient = mock(YarnClient.class);
+
+    Whitebox.setInternalState(cliUnderTest, "stramAgent", stramAgent);
+    Whitebox.setInternalState(cliUnderTest, "yarnClient", yarnClient);
+    Whitebox.setInternalState(cliUnderTest, "consolePresent", true);
+
+    suppress(constructor(WebServicesClient.class, new Class[0]));
+
+    List<ApplicationReport> runningApplications = new ArrayList<>();
+
+    ApplicationReport app1 = mockRunningApplicationReport("application-id-1", "app1");
+    ApplicationReport app2 = mockRunningApplicationReport("application-id-2", "app2");
+
+    runningApplications.add(app1);
+    runningApplications.add(app2);
+
+    when(yarnClient.getApplications(Mockito.any(Set.class))).thenReturn(runningApplications);
+    when(stramAgent.issueStramWebRequest(
+        Mockito.any(WebServicesClient.class),
+        Mockito.anyString(),
+        Mockito.any(StramAgent.StramUriSpec.class),
+        Mockito.any(WebServicesClient.WebServicesHandler.class)))
+        .thenReturn(new JSONObject());
+
+    final ByteArrayOutputStream stdOut = new ByteArrayOutputStream();
+    final ByteArrayOutputStream stdErr = new ByteArrayOutputStream();
+
+    PrintStream beforeOut = System.out;
+    PrintStream beforeErr = System.err;
+
+    System.setOut(new PrintStream(stdOut));
+    System.setErr(new PrintStream(stdErr));
+
+    // When processing the shutdown command for two valid and one invalid appNames
+    String shutdownAppsCommand = "shutdown-app app1 notExisting app2";
+    cliUnderTest.processLine(shutdownAppsCommand, new ConsoleReader(), true);
+
```

Comment:
Wouldn't it be a good idea to flush the current System.out and System.err (or close them) to ensure you get complete/consistent output in your stdOut and stdErr?